decisions/2020/03/10-ui-layers-stacking-context.md

```diff
@@ -0,0 +1,91 @@
+# 2020-03-10, UI: Layering and stacking
+
+## Attending
+
+Hendrik, Jan-Gerke, Joe, Ismay, Viktor
+
+## Context
+
+History and links to related issues and PRs: https://github.com/dhis2/ui-core/issues/596
+
+**TL;DR** is that we are not happy with the current solution and think
+we should do better. The current solution was the best we could
+implement given the pressure on getting a new major version out the door
+as it was blocking other development efforts. It serves its purpose but
+we suspect that there are limitations that we have not considered.
+
+Hendrik has created a presentation documenting the [evolution and
+current state of affairs + a new
+proposal](https://docs.google.com/presentation/d/1xCNPx08bT9uNtlBXt3zQRpV0xQwCoFTjjvJQd3e-K6g/edit#slide=id.g7f074dd6e5_0_92).
+
+## Proposals
+
+### New recursive layer + stacking context implementation
+
+https://github.com/dhis2/ui-core/pull/807
+
+### Current implementation
+
+https://github.com/dhis2/ui-core/pull/582
+
+Stick with the current implementation until known issues come up that
+force us to reimplement a different technique.
+
+The argument is that we have no known issues at the time of writing with
+the current solution.
+
+### Layer components implementation
+
+A completely different approach would be to avoid having to
+"reimplement" how CSS stacks layers and `z-index`:es by using components
+to define what should appear where.
+
+```
+const LayerBase = () => (....) // we won't need this one
+const LayerAlert = () => (....)
+const LayerError = () => (....)
+
+// index.html
+<body>
+  <div id="root" /> // basically LayerBase
+  <div id="layer-alert" /> // created by LayerAlert
+  <div id="layer-error /> // created by LayerError
+</body>
+
+// in the app
+<LayerAlert> // will render the children into #layer-alter
+  <Modal>
+    <Select />
+  </Modal>
+<LayerAlert>
+```
+
+## Conclusion
+
+- We find ourselves jumping into technical implementations before
+  resolving the actual needs that we have.
+
+- We need to take a step back to understand the needs and requirements for how our
+  libraries are supposed to work and what use cases we support in terms
+  of layers and popups.
+
+- Limitations needs to be documented and have documentation on how the
+  app should work to mitigate those shortcomings.
+
+- We need to define the behavior of how our layers should stack in feature
+  files.
```

Comment:
To clarify how I would see these feature files, I think that what would be helpful would be to have tests that compose our components the way an app would. So not testing our `<Layer>` components directly (or whichever component ends up doing the work), but very high level tests that combine our components the way lib users would, and checking the visibility.

If we make the tests too low level (so just on the level of a `<Layer>` component for example, we might still get stuck in technical discussions that have little relation to the actual problems our users might encounter.

Reply:
(After that we could always test the component directly as well of course. But for guiding the design discussion I would prefer the above.)

Reply:
@ismay I agree that testing things that way would be best, and I also think that's how @varl meant it yesterday. For testing the current implementation that approach should work well. 

However, if/when we get to testing my "improved" solution, this approach will be a bit too involved, because all these components would need to be refactored first so they use they use the `Layer` component. Perhaps, we can just quickly craft some dummy components that share the same layering characteristics as the components we tested in the current solution?

I guess the above ties in with your last suggestion:
> After that we could always test the component directly as well of course. But for guiding the design discussion I would prefer the above.

_BTW: I've reworked my "improved" solution and it's now much less complex. I think it still makes a lot of sense to switch to a layer system that leverages the z-index stacking context._

Reply:
I've added a couple of points in 3ba70fd that I hope address the concerns above.